service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/azure/container/CopyAzureStorageContainerBlobsStep.java

```diff
@@ -3,13 +3,90 @@
-// source container to the destination
 public class CopyAzureStorageContainerBlobsStep implements Step {
+  private final ResourceDao resourceDao;
+  private final ControlledAzureStorageContainerResource sourceContainer;
+  private final AzureStorageAccessService azureStorageAccessService;
+  private final AuthenticatedUserRequest userRequest;
+  private final BlobCopier blobCopier;
+
+  public CopyAzureStorageContainerBlobsStep(
+      AzureStorageAccessService azureStorageAccessService,
+      ControlledAzureStorageContainerResource sourceContainer,
+      ResourceDao resourceDao,
+      AuthenticatedUserRequest userRequest,
+      BlobCopier blobCopier) {
+    this.azureStorageAccessService = azureStorageAccessService;
+    this.resourceDao = resourceDao;
+    this.sourceContainer = sourceContainer;
+    this.userRequest = userRequest;
+    this.blobCopier = blobCopier;
+  }
+
   @Override
-  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    var inputParameters = flightContext.getInputParameters();
+
+    FlightUtils.validateRequiredEntries(
+        inputParameters, WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_WORKSPACE_ID);
+    FlightUtils.validateRequiredEntries(
+        flightContext.getWorkingMap(),
+        WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_STORAGE_ACCOUNT_RESOURCE_ID,
+        WorkspaceFlightMapKeys.ControlledResourceKeys.CLONED_RESOURCE_DEFINITION);
+
+    var destinationWorkspaceId =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+
+    var destStorageAccountResourceId =
+        flightContext
+            .getWorkingMap()
+            .get(
+                WorkspaceFlightMapKeys.ControlledResourceKeys
+                    .DESTINATION_STORAGE_ACCOUNT_RESOURCE_ID,
+                UUID.class);
+
+    ControlledAzureStorageResource destinationStorageAccount =
+        resourceDao
+            .getResource(destinationWorkspaceId, destStorageAccountResourceId)
+            .castByEnum(WsmResourceType.CONTROLLED_AZURE_STORAGE_ACCOUNT);
+
+    ControlledAzureStorageResource sourceStorageAccount =
+        resourceDao
+            .getResource(sourceContainer.getWorkspaceId(), sourceContainer.getStorageAccountId())
+            .castByEnum(WsmResourceType.CONTROLLED_AZURE_STORAGE_ACCOUNT);
+
+    ControlledAzureStorageContainerResource destinationContainer =
+        flightContext
+            .getWorkingMap()
+            .get(
+                WorkspaceFlightMapKeys.ControlledResourceKeys.CLONED_RESOURCE_DEFINITION,
+                ControlledAzureStorageContainerResource.class);
+
+    var results =
+        blobCopier.copyBlobs(
+            sourceStorageAccount, destinationStorageAccount, sourceContainer, destinationContainer);
```

Comment:
What happens if WSM needs to rerun this step? Will the blob copier be able to handle re-copying over the same files if some files have already been copied over on a previous run?

Reply:
Yes, the destination blob will be overwritten, see these [remarks](https://learn.microsoft.com/en-us/rest/api/storageservices/copy-blob#remarks).